src/systems/pose_publisher/PosePublisher.cc

```diff
@@ -277,38 +443,79 @@ void PosePublisherPrivate::InitializeEntitiesToPublish(
     }
   }
 
-  this->poses.reserve(this->entitiesToPublish.size());
+  // sanity check to make sure dynamicEntities are a subset of entitiesToPublish
+  for (auto const &ent : this->dynamicEntities)
+  {
+    if (this->entitiesToPublish.find(ent) == this->entitiesToPublish.end())
+    {
+      ignwarn << "Entity id: '" << ent << "' not found when creating a list "
+              << "of dynamic entities in pose publisher." << std::endl;
+    }
+  }
+
+  if (this->staticPosePublisher)
+  {
+    this->poses.reserve(this->dynamicEntities.size());
+    this->staticPoses.reserve(
+        this->entitiesToPublish.size() - this->dynamicEntities.size());
+  }
+  else
+  {
+    this->poses.reserve(this->entitiesToPublish.size());
+  }
 }
 
 //////////////////////////////////////////////////
-void PosePublisherPrivate::FillPoses(const EntityComponentManager &_ecm)
+void PosePublisherPrivate::FillPoses(const EntityComponentManager &_ecm,
+    std::vector<std::pair<Entity, math::Pose3d>> &_poses, bool _static)
 {
   IGN_PROFILE("PosePublisher::FillPose");
-  this->poses.clear();
+
   for (const auto &entity : this->entitiesToPublish)
   {
     auto pose = _ecm.Component<components::Pose>(entity.first);
-    if (pose)
-      this->poses.emplace_back(entity.first, pose->Data());
+    if (!pose)
+      continue;
+
+    bool isStatic = this->dynamicEntities.find(entity.first) ==
+          this->dynamicEntities.end();
+
+    if (_static == isStatic)
+      _poses.emplace_back(entity.first, pose->Data());
   }
 }
 
 //////////////////////////////////////////////////
-void PosePublisherPrivate::PublishPoses(const msgs::Time &_stampMsg)
+void PosePublisherPrivate::PublishPoses(
+    std::vector<std::pair<Entity, math::Pose3d>> &_poses,
+    const msgs::Time &_stampMsg,
+    transport::Node::Publisher &_publisher)
 {
   IGN_PROFILE("PosePublisher::PublishPoses");
+
   // publish poses
-  for (const auto &[entity, pose] : this->poses)
+  ignition::msgs::Pose *msg = nullptr;
+  if (this->usePoseV)
+    this->poseVMsg.Clear();
+  for (const auto &[entity, pose] : _poses)
   {
     auto entityIt = this->entitiesToPublish.find(entity);
     if (entityIt == this->entitiesToPublish.end())
       continue;
+
+    if (this->usePoseV)
+      msg = this->poseVMsg.add_pose();
+    else
+    {
+      this->poseMsg.Clear();
+      msg = &this->poseMsg;
+    }
+
     // fill pose msg
-    this->poseMsg.Clear();
     // frame_id: parent entity name
     // child_frame_id = entity name
     // pose is the transform from frame_id to child_frame_id
-    auto header = this->poseMsg.mutable_header();
+    auto header = msg->mutable_header();
```

Comment:
Is there any chance of `msg` being null here?

Reply:
it should never be null. Added `IGN_ASSERT` to be sure. 4255f22